Sync WIT with upstream specification, `use`, packages, and all

alexcrichton opened the pull request:
I'm opening this up now to share some early work on this. This is unfortunately going to again be a very large PR due to the sheer amount of infrastructure required to pull in all the WIT updates, namely to implement https://github.com/WebAssembly/component-model/pull/141. By-and-large the headline features of this PR will be:

* [x] The `use` keyword will be fully implemented. This means that WIT now has the concept of packages, dependencies between packages, documents, etc. Packages can be split between multiple files and packages can depend on other packages as well. All of the `use` statements are preserved in the internal representation to enable code generators to see connections between types and generate appropriate code, e.g. one `struct` in Rust.
* [x] A `wit-parser`-specific convention is invented here for a set of packages residing on the filesystem, intended to probably be in a `wit/` folder at the top of a project. The "root package" is located at `wit/*.wit` and dependencies are located at `wit/deps/$dep/*.wit`. Dependencies can use amongst themselves and this should provide a means by which to depend on packages (like WASI) before registry tooling is in place. Submodules may also help here too perhaps.
* [x] Full support in `wit-component` for all of these features. This means that a component can be created from a world which is itself assembled from many packages/files/interfaces/etc.
* [x] The `wasm-tools component new --types-only` interface is removed in favor of `wasm-tools component wit --binary`. More-or-less the binary format for WIT is now supported under `wasm-tools component wit` which can be used for printing the WIT of a preexisting component, converting a WIT binary to WIT text, or converting WIT text to WIT binary.

I'm opening this up as a work-in-progress at this time. I'm almost done with `wit-parser` and the text syntax bits. A good way to explore this is through the tests updated and added in this PR. Procedually I'll be taking next week and the week after off so this is not going to merge and get published until at least January 2023.

---

I'll preemptively apologize again for the size and scale of this PR. As with https://github.com/bytecodealliance/wasm-tools/pull/858 there's not really a great way to review this. That being said I think there's a case to be made for the WIT format being so "early days" that a meticulous review probably isn't worth it at this point anyway. We've got a process in place at the component-model repository to discuss syntax, semantics, etc, and the goal of this PR is for this repository to become a faithful implementation of that specification. Any surprising behavior should either be clarified upstream or fixed in this repository to match what's specified upstream. In that sense most of the "review" I think is best done as discussion about the concepts on PRs like https://github.com/WebAssembly/component-model/pull/141. The technical implementation here certainly has quite a lot to review as well, but personally I feel it's best to largely get everything working first to have a basis on which to iterate, as currently that doesn't altogether exist.

alexcrichton commented:
Ok all the tests pass, phew! Pieces remaining are:

* Type declarations in `world` blocks
* The `use` statement in `world` blocks
* Cleaning up with comments and TODOs and panics
* Extracting a `*.wit` package from a preexisting component (this is a big item I forgot until I just wrote this down)

at the very least though progress is being made!

alexcrichton commented:
Ok I think that this is ready to go. This is still missing types-in-`world` blocks but I think at this point it might be best to leave that to a follow-up PR.

alexcrichton commented:
https://github.com/bytecodealliance/wasmtime/pull/5565 is an example of updating a code generator for these changes.

peterhuene commented on crates/wit-component/tests/interfaces/diamond-disambiguate/shared1.wit.print:
```diff
@@ -0,0 +1,5 @@
+interface shared {
+  type t1 = u8
+
```
Note: don't need to fix this in this PR as it is a preexisting issue with the wit printer, but we should probably do a pass on the newlines it prints to make the output more consistent; perfect for a follow-up PR.

peterhuene commented on crates/wit-component/tests/interfaces/diamond-disambiguate/shared1.wit.print:
```diff
@@ -0,0 +1,5 @@
+interface shared {
```
Do you think a future improvement could recover from the encoding the fact that this interface is `default` for this document?

alexcrichton replied:
I'd like to, yeah, but in talking with Luke we haven't figure out a great way to encode it in wasm just yet.

peterhuene reviewed (commented):
Review status: I'm about 60% of the way through the files after completing everything under `wit-component` (including test updates).

Submitting the comments I've had up to this point.

peterhuene commented:
@alexcrichton something I noticed locally: it looks like `crates/wit-component/fuzz/fuzz_targets/roundtrip-wit.rs` needs updating.

alexcrichton commented on crates/wit-component/src/decoding.rs:
```diff
@@ -189,420 +286,632 @@ struct InterfaceDecoder<'a, 'doc> {
+                        Ok(Field {
+                            name: name.to_string(),
+                            ty: self.convert_valtype(ty)?,
+                            docs: Default::default(),
```
I definitely agree, and it's not something I've thought too hard about yet. I think we'll want something akin to the `name` custom section for the component model, but when I first thought of that I realized that the `name` section is not suitable for naming items within namespaces such as types, which is where the entire encoding of a WIT package resides.

Basically I think we need to solve the problem of the `name` section being able to name sub-namespace items (such as nested types within an instance type in the type section) and then model a documentation custom section on that same format. I'm not 100% sure how best to get all the info from wasmparser to here but I'm hoping it won't be too too hard.

alexcrichton commented on crates/wit-component/tests/components/adapt-preview1/module.wat:
```diff
@@ -3,8 +3,8 @@
   ;; import some wasi functions
-  (import "wasi_snapshot_preview1" "proc_exit" (func (param i32)))
-  (import "wasi_snapshot_preview1" "random_get" (func (param i32 i32) (result i32)))
+  (import "wasi-snapshot-preview1" "proc_exit" (func (param i32)))
```
Yeah I tried to figure out a good way here but I concluded in the end it wasn't worth it. This doesn't really have anything to do with WASI itself but moreso the conventions of the tests here where `foo.wat` has a corresponding `foo.wit` and the `foo` part of the filestem has to be a valid WIT identifier because it becomes the document name.

Currently I haven't added various APIs to `UnresolvedPackage::parse` to handle a document/package name different from the filename itself, so while that could be done I figured it may not be too worth it as this is just a test-convention thing rather than a problem-with-WASI thing.

alexcrichton commented on crates/wit-component/tests/components/lift-options/module.wit:
```diff
@@ -33,6 +33,6 @@ interface my-default {
-world my-world {
-  default export my-default
+default world my-world {
+  export foo: self.my-default
```
Agreed! Initially when I talked with Luke about this he thought it would be best to leave out the splat syntax for now, but given that this PR doesn't yet fully implement types-in-worlds there's no true replacement for `default export` just yet.

I'm curious, do you think that releasing these changes and/or merging this PR should be gated on figuring this out? I could go either way myself.

peterhuene commented on crates/wit-parser/src/ast/resolve.rs:
```diff
@@ -30,245 +53,682 @@ enum Key {
+                    }),
+                    None => bail!(Error {
+                        span: name.span,
+                        msg: format!("name is not defined"),
```
```suggestion
                        msg: format!("name `{name}` is not defined", name = name.name),
```

peterhuene commented on crates/wit-parser/src/ast/resolve.rs:
```diff
@@ -30,245 +53,682 @@ enum Key {
+                    Some(InterfaceItem::Type(id)) => *id,
+                    Some(InterfaceItem::Func) => bail!(Error {
+                        span: name.span,
+                        msg: format!("cannot use a function as a type"),
```
```suggestion
                        msg: format!("cannot use function `{name}` as a type", name = name.name),
```

peterhuene commented on src/bin/wasm-tools/component.rs:
```diff
@@ -71,19 +80,66 @@ pub struct NewOpts {
+
+        let bytes = encoder
+            .encode()
+            .with_context(|| format!("failed to encode a component from module "))?;
```
```suggestion
            .context("failed to encode a component from module")?;
```

peterhuene commented on src/bin/wasm-tools/component.rs:
```diff
@@ -92,96 +148,347 @@ pub struct NewOpts {
+    /// text form.
+    ///
+    /// This is incompatible with `-o`.
+    #[clap(long)]
```
```suggestion
    #[clap(long, conflicts_with = "output", conflicts_with = "wasm", conflicts_with = "wat", conflicts_with = "document")]
```

peterhuene commented on src/bin/wasm-tools/component.rs:
```diff
@@ -92,96 +148,347 @@ pub struct NewOpts {
+    #[clap(short, long)]
+    wasm: bool,
+
+    /// When combined with `--wasm` this will print the WebAssembly text format
+    /// instead of the WebAssembly binary format.
+    #[clap(short = 't', long)]
```
```suggestion
    /// Emit a WebAssembly textual representation instead of the WIT text format.
    #[clap(short = 't', long, conflicts_with = "wasm")]
```

peterhuene commented on src/bin/wasm-tools/component.rs:
```diff
@@ -92,96 +148,347 @@ pub struct NewOpts {
+    #[clap(long)]
+    out_dir: Option<PathBuf>,
+
+    /// Emit the WIT binary format, a WebAssembly component, instead of the WIT
+    /// text format.
+    #[clap(short, long)]
```
I think the `--wasm` and `--wat` options should be mutually exclusive rather than the additive. Requiring `--wat` to need `--wasm` is a little 

Thus the default output is `wit`, `--wasm` generates a wasm binary encoding, and `--wat` generates a wasm text encoding.

```suggestion
    /// Emit a WebAssembly binary representation instead of the WIT text format.
    #[clap(short, long, conflicts_with = "wat")]
```

peterhuene commented on src/bin/wasm-tools/component.rs:
```diff
@@ -92,96 +148,347 @@ pub struct NewOpts {
+    #[clap(short = 't', long)]
+    wat: bool,
+
+    /// When specifying `--wasm` the output wasm binary is validated by default,
+    /// and this can be used to skip that step.
```
```suggestion
    /// Skips the validation performed when using the `--wasm` and `--wat` options.
```

peterhuene commented on src/bin/wasm-tools/component.rs:
```diff
@@ -92,96 +148,347 @@ pub struct NewOpts {
+        // This interprets all of the output options and performs such a task.
+        match &self.out_dir {
+            Some(dir) => {
+                if self.output.output_path().is_some() {
+                    bail!("cannot specify both `--out-dir` and `--output`");
+                }
+                if self.wasm {
+                    bail!("cannot specify both `--out-dir` and `--wasm`");
+                }
+                if self.wat {
+                    bail!("cannot specify both `--out-dir` and `--wat`");
+                }
+                if self.document.is_some() {
+                    bail!("cannot specify both `--out-dir` and `--document`");
+                }
```
```suggestion
                assert!(self.output.output_path().is_none());
                assert!(!self.wasm && !self.wat);
                assert!(self.document.is_none());
```

peterhuene commented on src/bin/wasm-tools/component.rs:
```diff
@@ -92,96 +148,347 @@ pub struct NewOpts {
+                    }
+                };
+                let resolve = decoded.resolve();
+                std::fs::create_dir_all(&dir).context(format!("failed to create {dir:?}"))?;
```
```suggestion
                std::fs::create_dir_all(&dir).with_context(|| format!("failed to create {dir:?}"))?;
```

peterhuene commented on src/bin/wasm-tools/component.rs:
```diff
@@ -92,96 +148,347 @@ pub struct NewOpts {
+                for (name, doc) in resolve.packages[package].documents.iter() {
+                    let output = DocumentPrinter::default().print(&resolve, *doc)?;
+                    let path = dir.join(format!("{name}.wit"));
+                    std::fs::write(&path, output).context(format!("failed to write {path:?}"))?;
```
```suggestion
                    std::fs::write(&path, output).with_context(|| format!("failed to write {path:?}"))?;
```

peterhuene commented on src/bin/wasm-tools/component.rs:
```diff
@@ -92,96 +148,347 @@ pub struct NewOpts {
+                }
+            }
+            None => {
+                if self.wasm {
```
```suggestion
                if self.wasm || self.wat {
```

peterhuene commented on src/bin/wasm-tools/component.rs:
```diff
@@ -92,96 +148,347 @@ pub struct NewOpts {
+    fn emit_wit(&self, decoded: &DecodedWasm) -> Result<()> {
+        assert!(!self.wasm);
+        assert!(self.out_dir.is_none());
+        if self.wat {
+            bail!("the `--wat` option can only be combined with `--wasm`");
+        }
```
```suggestion
```

peterhuene commented on src/bin/wasm-tools/component.rs:
```diff
@@ -92,96 +148,347 @@ pub struct NewOpts {
+    }
+
+    fn emit_wit(&self, decoded: &DecodedWasm) -> Result<()> {
+        assert!(!self.wasm);
```
```suggestion
        assert!(!self.wasm && !self.wat);
```

peterhuene commented on src/bin/wasm-tools/component.rs:
```diff
@@ -92,96 +148,347 @@ pub struct NewOpts {
     }
+
+    fn emit_wasm(&self, decoded: &DecodedWasm) -> Result<()> {
+        assert!(self.wasm);
```
```suggestion
        assert!(self.wasm || self.wat);
```

peterhuene commented on src/bin/wasm-tools/component.rs:
```diff
@@ -92,96 +148,347 @@ pub struct NewOpts {
+    /// When printing a WIT package, the default mode, this option is used to
+    /// indicate which document is printed within the package if more than one
+    /// document is present.
+    #[clap(short, long)]
```
```suggestion
    #[clap(short, long, conflicts_with = "wasm", conflicts_with = "wat")]
```

peterhuene commented on src/bin/wasm-tools/component.rs:
```diff
@@ -92,96 +148,347 @@ pub struct NewOpts {
+    fn emit_wasm(&self, decoded: &DecodedWasm) -> Result<()> {
+        assert!(self.wasm);
+        assert!(self.out_dir.is_none());
+        if self.document.is_some() {
+            bail!("cannot specify `--document` with `--wasm`");
+        }
```
```suggestion
        assert!(self.document.is_none());
```

peterhuene reviewed (approved):
I played around with the CLI tooling changes locally and reviewed each file of the PR.

Overall, the changes look great and I'm excited to start building stuff on top of this as wit now becomes a lot more flexible.

The feedback is mostly nits, except for a minor DX change to have the `--wat` option for the `component wit` command  not need the `--wasm` option.

Marking as approved as we don't absolutely have to address the feedback to land this.

peterhuene commented:
Apologies for all the suggestions to the `component wit` command as I was using the PR review tools as a code editor, apparently 🙃.

Feel free to ignore/resolve them all if you'd rather keep it as-is or implement it differently from the suggestions.

alexcrichton commented on crates/wit-parser/src/live.rs:
```diff
@@ -0,0 +1,120 @@
+        for ty in func.results.iter_types() {
+            self.add_type(resolve, ty);
+        }
+        // ..
```
Ah no accidentally left this in

alexcrichton commented:
Thanks for the thorough review! I'm putting the finishing touches on the fuzzer but have otherwise pushed up addressed feedback plus bugs the fuzzer has found so far.

alexcrichton commented:
Ok I've run the fuzzer long enough locally that the bugs are becoming more obscure that I think it's time to get this in. @peterhuene mind double-checking the commits I added since your last review?

Thanks again for taking the time to go through all of this!

peterhuene commented on fuzz/fuzz_targets/roundtrip-wit.rs:
```diff
@@ -0,0 +1,719 @@
+        fn gen(&mut self, u: &mut Unstructured<'_>) -> Result<Vec<Package>> {
+            let mut packages = Vec::new();
+            let mut names = HashSet::new();
+            while packages.len() < 10 && (packages.is_empty() || u.arbitrary()?) {
```
Nit: can we add constants for these maximums (here and in a few other places in c42cfd2ab471e20dc78812131e05c3211d5e474a) to go along side `MAX_PARTS`? It just makes it a little easier to understand that these are limits being placed intentionally on what is generated.

peterhuene reviewed (approved):
Looks great! Just one last nit comment related to the new fuzz target.
